internals/overlord/state/notices.go

```diff
@@ -0,0 +1,386 @@
+// Copyright (c) 2023 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package state
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"sort"
+	"strconv"
+	"time"
+
+	"github.com/canonical/pebble/internals/logger"
+)
+
+const (
+	MaxNoticeKeyLength = 255
+
+	noticeExpireAfter = 7 * 24 * time.Hour
+)
+
+// Notice represents an aggregated notice. The combination of type and key is unique.
+type Notice struct {
+	// Server-generated unique ID for this notice (a surrogate key). Users
+	// shouldn't rely on this, but this will be a monotonically increasing
+	// number (like change ID).
+	id string
+
+	// The notice's type.
+	noticeType NoticeType
+
+	// The notice key: a string that must be unique for this notice type.
+	// For example, for the "change-update" notice, the key is the change ID.
+	// For a warning, the key would be the warning message.
+	//
+	// This is limited to a maximum of 255 bytes when added (it's an error
+	// to add a notice with a longer key).
+	key string
+
+	// The first time one of these notices (type and key combination) occurred.
+	firstOccurred time.Time
+
+	// The last time one of these notices occurred.
+	lastOccurred time.Time
+
+	// Same as lastOccurred when the notice was last repeated. Only updated
+	// once lastOccurred > lastRepeated + repeatAfter.
+	lastRepeated time.Time
```

Comment:
First, I agree that the terms "repeat" and "last-repeated" aren't obvious at first -- I actually struggled with them when we were working on the spec too. However, we brainstormed quite a bit and couldn't come up with an obviously better term, so "last repeated" and "repeat after" stuck.

HOWEVER, with this comment and many of those below, I think you're misunderstanding how this works. Unlike warnings, we don't update the notice when it was "announced". The Warnings design would be problematic (for Notices) because it can only have a single reader and actually updates the `lastShown` time when it was shown. But notices is intended to be able to have multiple readers, each of which can keep its own "most recent" time to use in the `After` query field -- so the "what we've last shown" or "what we've last announced" state is stored by the client.

I've tried to clarify the wording on several comments to explain this better, especially the comment on the `lastRepeated` field. Let me know if this helps explain it, or if you're still confused.

Because many of the following comments are related to this confusion, I've resolved similar comments below so we can keep the discussion in one place.

Reply:
All good! Thank you.